File: src/java/com/distributedlife/mahjong/reference/hand/HandLibraryBuilder.java
package com.distributedlife.mahjong.reference.hand;

import com.distributedlife.mahjong.reference.adapter.HandCandidateToAHandConverter;
import com.distributedlife.mahjong.reference.data.TileSet;
import com.distributedlife.mahjong.reference.filter.HandCandidateFilter;
import com.distributedlife.mahjong.reference.permute.PermutatorExecutor;

import java.util.*;

public class HandLibraryBuilder {
    private List<HandCandidateFilter> filters;
    private final PermutatorExecutor permutatorExecutor;
    private final TileSet tileSet;
    private final List<HandDefinition> definitions;
    private HandCandidateToAHandConverter converter;

    public HandLibraryBuilder(TileSet tileSet, List<HandDefinition> definitions, List<HandCandidateFilter> filters, PermutatorExecutor permutatorExecutor, HandCandidateToAHandConverter converter) {
        this.tileSet = tileSet;
        this.definitions = definitions;
        this.filters = filters;
        this.permutatorExecutor = permutatorExecutor;
        this.converter = converter;
    }

    public List<Hand> buildAll() {
        List<HandCandidate> candidates = new ArrayList<HandCandidate>();

        for (HandDefinition definition : definitions) {
            candidates.addAll(build(definition));
        }

        for (HandCandidateFilter filter : filters) {
            candidates = filter.apply(candidates);
        }

        List<Hand> handsAsList = runListThroughSetToEnsureUnique(new HashSet<Hand>(converter.convert(candidates)));

        Collections.sort(handsAsList);

        return handsAsList;
    }

    private List<Hand> runListThroughSetToEnsureUnique(Set<Hand> handsAsSet) {
        List<Hand> handsAsList = new ArrayList<Hand>();
        for(Hand hand : handsAsSet) {
            handsAsList.add(hand);
        }
        return handsAsList;
    }

    private List<HandCandidate> build(HandDefinition definition) {
        List<HandCandidate> candidates = new ArrayList<HandCandidate>();

        for (String suit : definition.getSuits()) {
            HandCandidate candidate = new HandCandidate(definition.getName(), new ArrayList<String>(tileSet.getTiles()));
            candidate.setPrimarySuit(suit);

            candidates.addAll(permutatorExecutor.runPermutatorsOnCandidate(candidate, definition.getRequirements()));
        }

        return candidates;
    }
}

File: lib/hand/hand-library-builder.js
/**
 * HandLibraryBuilder - Generates all valid tile combinations for defined hands
 */
import { HandCandidate } from './hand-candidate.js';

export class HandLibraryBuilder {
  constructor(tileSet, definitions, filters, permutatorExecutor, converter) {
    this.tileSet = tileSet;
    this.definitions = definitions;
    this.filters = filters;
    this.permutatorExecutor = permutatorExecutor;
    this.converter = converter;
  }

  async buildAll() {
    let candidates = [];

    for (const definition of this.definitions) {
      const handCandidates = await this.build(definition);
      candidates = candidates.concat(handCandidates);
    }

    for (const filter of this.filters) {
      candidates = filter.apply(candidates);
    }

    let hands = this.converter.convert(candidates);
    
    // Ensure unique hands
    const uniqueHands = new Map();
    for (const hand of hands) {
      const key = `${hand.name}|${hand.part1}|${hand.part2}|${hand.part3}|${hand.part4}`;
      uniqueHands.set(key, hand);
    }
    
    hands = Array.from(uniqueHands.values());
    
    // Sort hands
    hands.sort((a, b) => a.compareTo(b));

    return hands;
  }

  async build(definition) {
    const candidates = [];

    for (const suit of definition.getSuits()) {
      const candidate = new HandCandidate(
        definition.getName(),
        [...this.tileSet.getTiles()]
      );
      candidate.setPrimarySuit(suit);

      const results = this.permutatorExecutor.runPermutatorsOnCandidate(
        candidate,
        definition.getRequirements()
      );
      candidates.push(...results);
    }

    return candidates;
  }
}
